OptimizelySDK/Odp/OdpEventManager.cs

```diff
@@ -0,0 +1,448 @@
+            }
+
+            _logger.Log(LogLevel.DEBUG, $"Flushing Queue.");
+
+            CurrentState = ExecutionState.Processing;
+            while (QueueContainsItems())
+            {
+                DequeueSendSingleBatch();
+            }
+
+            CurrentState = ExecutionState.Running;
+        }
+
+        /// <summary>
+        /// Dequeue a single batch and send it
+        /// </summary>
+        private void DequeueSendSingleBatch()
+        {
+            var batch = new List<OdpEvent>(_batchSize);
+
+            // dequeue a batch 
+            for (int i = 0; i < _batchSize && _queue.Count > 0; i += 1)
+            {
+                if (_queue.TryDequeue(out OdpEvent dequeuedOdpEvent))
+                {
+                    batch.Add(dequeuedOdpEvent);
+                }
+            }
+
+            if (batch.Count > 0)
+            {
+                Task.Run(() =>
+                {
+                    bool shouldRetry;
+                    var attemptNumber = 0;
+                    do
+                    {
+                        shouldRetry = _odpEventApiManager.SendEvents(_odpConfig.ApiKey,
+                            _odpConfig.ApiHost, batch);
+                        attemptNumber += 1;
+                    } while (shouldRetry && attemptNumber < MAX_RETRIES);
+                });
+            }
+        }
+
+        /// <summary>
+        /// Determines if the ODP configuration is ready
+        /// </summary>
+        /// <returns>True if required parameters are available otherwise False</returns>
+        private bool IsOdpConfigurationReady()
+        {
+            if (_odpConfig.IsReady())
+            {
+                return true;
+            }
+
+            _logger.Log(LogLevel.DEBUG, "ODP is not integrated.");
+
+            // ensure empty queue
+            _queue = new ConcurrentQueue<OdpEvent>();
+
+            return false;
+        }
+
+        /// <summary>
+        /// Queue count has enough items to send at least one batch
+        /// </summary>
+        /// <returns>True if even batches exist otherwise False</returns>
+        private bool QueueHasBatches()
+        {
+            return QueueContainsItems() && _queue.Count >= _batchSize;
+        }
+
+        /// <summary>
+        /// Queue contains a items
+        /// </summary>
+        /// <returns>True if count is > 0 otherwise False</returns>
+        private bool QueueContainsItems()
+        {
+            return _queue.Count > 0;
+        }
+
+        /// <summary>
+        /// ODP event data has invalid types
+        /// </summary>
+        /// <param name="data">Data to be analyzed</param>
+        /// <returns>True if a type is not a valid type or is not null otherwise False</returns>
+        private bool InvalidDataFound(Dictionary<string, dynamic> data)
+        {
+            return data.Any(item =>
+                item.Value != null &&
+                !_validOdpDataTypes.Contains(item.Value.GetType().Name)
```

Comment:
should have ignore case.

Reply:
Weird. `.Contains()` didn't have a second parameter. Solving alternatively